### PR TITLE
checker: disallow comparison between enum and int

### PR DESCRIPTION
**Pull request by Delta456**
Closes https://github.com/vlang/v/issues/7512

### CODE REVIEW

File: vlib/v/checker/tests/compare_enum_and_int_err.out [hunk 1]
```diff
@@ -0,0 +1,6 @@
+vlib/v/checker/tests/compare_enum_and_int_err.vv:7:16: error: cannot compare an `int` and `enum`, use explicit type cast for `int` or `enum` instead
```
Comment (Delta456) [inline vlib/v/checker/tests/compare_enum_and_int_err.out:1]: Does `with` sounds better than `and`?

**Comment by Delta456** [comment]
Now I think about should we make this a warning for now as you can see that many codebases use it right now. What you think @medvednikov?

**Comment by UweKrueger** [comment]
@Delta456 There are these lines in `check_types.v` function `basic_check()`:
```v
// allow enum value to be used as int
if (got_type_sym.is_int() && exp_type_sym.kind == .enum_) ||
    (exp_type_sym.is_int() && got_type_sym.kind == .enum_) {
    return true
}
```
Have you considered removing them, instead of introducing an additional check? It might simplify the type system and make it more consistent.
Either `enum` and `int` should be compatible everywhere or nowhere.

**Comment by Delta456** [comment]
> 
> 
> @Delta456 There are these lines in `check_types.v` function `basic_check()`:
> 
> ```v
> // allow enum value to be used as int
> if (got_type_sym.is_int() && exp_type_sym.kind == .enum_) ||
>     (exp_type_sym.is_int() && got_type_sym.kind == .enum_) {
>     return true
> }
> ```
> 
> Have you considered removing them, instead of introducing an additional check? It might simplify the type system and make it more consistent.
> Either `enum` and `int` should be compatible everywhere or nowhere.

I was finding this only, thanks! I think we should allow it for now and give a warning instead of an error as it is used a lot in many codebases.

**Comment by UweKrueger** [comment]
> ... I think we should allow it for now and give a warning instead of an error as it is used a lot in many codebases.

I don't know...  Consider this:
```v
color = 5           // allowed
if color == 5 { ... // forbidden
```
And postponing breaking changes to after 1.0 is not a really good strategy ;-)

File: vlib/v/checker/tests/compare_enum_and_int_err.out (hunk 1 above)
Comment (danieldaeschle) [reply vlib/v/checker/tests/compare_enum_and_int_err.out:1]: deepl.com says `with`

Comment (JalonSolov) [reply vlib/v/checker/tests/compare_enum_and_int_err.out:1]: So do I.  :-)

**Comment by medvednikov** [comment]
Neither assigning numbers to enums, nor comparing them to numbers without a cast can be allowed.

File: vlib/v/checker/checker.v [hunk 2]
```diff
@@ -2412,6 +2412,13 @@ pub fn (mut c Checker) assign_stmt(mut assign_stmt ast.AssignStmt) {
 			// TODO replace all c.pref.translated checks with `$if !translated` for performance
 			continue
 		}
+		if ((left_sym.kind == .enum_ &&
+			right_sym.kind in [.any_int, .int]) ||
```
Comment (danieldaeschle) [inline vlib/v/checker/checker.v]: I think there is a is_number or is_int

File: vlib/v/checker/checker.v [hunk 3]
```diff
@@ -2412,6 +2412,13 @@ pub fn (mut c Checker) assign_stmt(mut assign_stmt ast.AssignStmt) {
...
+			(right_sym.kind == .enum_ && left_sym.kind in [.any_int, .int])) &&
+			!is_decl {
+			c.warn('cannot assign `enum` to `int` and vice versa, use explicit type cast for `int` or `enum` instead',
```
Comment (danieldaeschle) [inline vlib/v/checker/checker.v:2419]: Please detect the direction a d print the message correctly without vice versa. Just use sym.name

File: vlib/v/checker/checker.v (hunk 2 above)
Comment (Delta456) [reply vlib/v/checker/checker.v]: But `enum` only allows to be 2^32 -1 values so that check isn't needed here.

Comment (danieldaeschle) [reply vlib/v/checker/checker.v]: I mean, it‘s a easier check to check if it‘s a int.

File: vlib/v/checker/check_types.v [hunk 4]
```diff
@@ -89,6 +89,7 @@ pub fn (mut c Checker) check_basic(got table.Type, expected table.Type) bool {
 		return true
 	}
 	// allow enum value to be used as int
+	// TODO: remove this after 0.3 release and the related warning to it
 	if (got_type_sym.is_int() && exp_type_sym.kind == .enum_) ||
 		(exp_type_sym.is_int() && got_type_sym.kind == .enum_) {
```
Comment (UweKrueger) [inline vlib/v/checker/check_types.v:94]: It might be better to put the warning here instead of handling comparison and assignment special.
There are also other expressions (`enum` + `int`, ...) that should not be allowed. Putting the warning here shows us now what will happen when we delete this exception.

Comment (Delta456) [reply vlib/v/checker/check_types.v:94]: I think the current way is fine because this will be removed after 0.3 anyways else it will just become more complex IMHO.

Comment (UweKrueger) [reply vlib/v/checker/check_types.v:94]: Agreed. It would require too many changes for a temporary solution.

File: vlib/v/checker/checker.v [hunk 5]
```diff
@@ -932,6 +932,12 @@ pub fn (mut c Checker) infix_expr(mut infix_expr ast.InfixExpr) table.Type {
 		// TODO broken !in
 		c.error('string types only have the following operators defined: `==`, `!=`, `<`, `>`, `<=`, `>=`, and `+`',
 			infix_expr.pos)
+	} else if ((left.kind == .enum_ &&
+		right.kind in [.int, .any_int]) || (left.kind in [.int, .any_int] &&
```
Comment (UweKrueger) [inline vlib/v/checker/checker.v:936]: I agree with @danieldaeschle the check should be `right.is_int()` or `right_type.is_int()` and `left_type.is_int()`. So if `i64`, `i16` is used, we get a warning, too.

Comment (Delta456) [reply vlib/v/checker/checker.v:936]: Alright

Comment (Delta456) [reply vlib/v/checker/checker.v:936]: Done

File: vlib/v/checker/checker.v (hunk 3 above)
Comment (Delta456) [reply vlib/v/checker/checker.v]: Done

File: vlib/v/checker/checker.v (hunk 2 above)
Comment (Delta456) [reply vlib/v/checker/checker.v]: Done

**Review by UweKrueger — APPROVED** [review]
Looks OK to me. Can be merged.

**Review by danieldaeschle — CHANGES_REQUESTED** [review]
+ Add changelog statement (deprecated).
+ Add checker tests

File: vlib/v/checker/tests/enum_as_int_err.out [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+   10 |     foo = Color.red
+   11 |     println(color == 0)
+vlib/v/checker/tests/enum_as_int_err.vv:10:6: error: cannot assign `Color` to `int literal`, use explicit type cast for `Color` instead
+    8 |     mut foo := 1
+    9 |     color = 1
+   10 |     foo = Color.red
+      |         ^
+   11 |     println(color == 0)
+   12 | }
+vlib/v/checker/tests/enum_as_int_err.vv:11:16: error: cannot compare an `int` with `enum`, use explicit type cast for `int` or `enum` instead
```
Comment (danieldaeschle) [inline vlib/v/checker/tests/enum_as_int_err.out:15]: looks like hardcoded `int` and `enum`, please update this.

File: vlib/v/checker/tests/enum_as_int_err.out [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    7 |     mut color := Color.red
+    8 |     mut foo := 1
+    9 |     color = 1
+      |           ^
+   10 |     foo = Color.red
+   11 |     println(color == 0)
+vlib/v/checker/tests/enum_as_int_err.vv:10:6: error: cannot assign `Color` to `int literal`, use explicit type cast for `Color` instead
+    8 |     mut foo := 1
+    9 |     color = 1
+   10 |     foo = Color.red
```
Comment (danieldaeschle) [inline vlib/v/checker/tests/enum_as_int_err.out:11]: I'm not sure if it's correct to underline the `=`. How does similar errors look like? The error is the value on the right of `=` not the `=` itself. Maybe we should adjust it everywhere. But it can be another PR.

File: vlib/v/checker/checker.v [hunk 8]
```diff
@@ -936,7 +936,8 @@ pub fn (mut c Checker) infix_expr(mut infix_expr ast.InfixExpr) table.Type {
 	} else if ((left.kind == .enum_ && right.is_int()) ||
 		(left.is_int() && right.kind == .enum_)) &&
 		infix_expr.op.is_relational() {
-		c.warn('cannot compare an `int` with `enum`, use explicit type cast for `int` or `enum` instead',
+		name := if left.kind == .enum_ { left.name } else { right.name }
```
Comment (danieldaeschle) [inline vlib/v/checker/checker.v:939]: please use this:
```v
enum_name := if left.kind == .enum_ { left.name } else { right.name }
int_name := if left.kind == .enum_ { right.name } else { left.name }
c.warn('cannot compare an `$int_name ` with `$enum_name`, use explicit type cast instead',
			infix_expr.pos)
```

Comment (danieldaeschle) [reply vlib/v/checker/checker.v:939]: hardcoded `int literal` is bad practice.

**Review by danieldaeschle — CHANGES_REQUESTED** [review]

File: vlib/v/checker/checker.v [hunk 9]
```diff
@@ -936,8 +936,9 @@ pub fn (mut c Checker) infix_expr(mut infix_expr ast.InfixExpr) table.Type {
 	} else if ((left.kind == .enum_ && right.is_int()) ||
 		(left.is_int() && right.kind == .enum_)) &&
 		infix_expr.op.is_relational() {
-		name := if left.kind == .enum_ { left.name } else { right.name }
-		c.warn('cannot compare an `int literal` with `$name`, use explicit type cast for `int` or `$name` instead',
+		enum_name := if left.kind == .enum_ { left.name } else { right.name }
+		int_name := if left.kind == .enum_ { right.name } else { left.name }
+		c.warn('cannot compare an `$int_name ` with `$enum_name`, use explicit type cast instead',
```
Comment (danieldaeschle) [inline vlib/v/checker/checker.v:941]: that space is not intended i think

**Review by danieldaeschle — CHANGES_REQUESTED** [review]

File: vlib/v/checker/checker.v (hunk 9 above)
Comment (Delta456) [reply vlib/v/checker/checker.v:941]: this is from the formatter.

**Review by danieldaeschle — APPROVED** [review]

File: vlib/v/checker/checker.v [hunk 10]
```diff
@@ -936,7 +936,8 @@ pub fn (mut c Checker) infix_expr(mut infix_expr ast.InfixExpr) table.Type {
 	} else if ((left.kind == .enum_ && right.is_int()) ||
 		(left.is_int() && right.kind == .enum_)) &&
 		infix_expr.op.is_relational() {
-		c.warn('cannot compare an `int` with `enum`, use explicit type cast for `int` or `enum` instead',
+		name := if left.kind == .enum_ { left.name } else { right.name }
+		c.warn('cannot compare an `int literal` with `$name`, use explicit type cast for `int` or `$name` instead',
```
[the remark this answers is not in the document]
Comment (Delta456) [reply vlib/v/checker/checker.v]: Right 🙂

[the remark this answers is not in the document]
Comment (danieldaeschle) [reply vlib/v/checker/checker.v]: He has something similar right now?

**Comment by Delta456** [comment]
Can this be merged now?

File: examples/2048/2048.v [hunk 11]
```diff
@@ -797,7 +797,7 @@ fn (mut app App) next_theme() {
 
 [inline]
 fn (mut app App) next_tile_format() {
-	app.tile_format = int(app.tile_format) + 1
+	app.tile_format = app.tile_format + TileFormat(1)
```
Comment (spaceface777) [inline examples/2048/2048.v:800]: btw, I don't think math operators should work on enums directly, the correct thing to do here would be `TileFormat(int(app.tile_format) + 1)`.

This is way out of scope for this PR, just wanted to document it :-)

Comment (Delta456) [reply examples/2048/2048.v:800]: Ok, I will change it.

Comment (Delta456) [reply examples/2048/2048.v]: Done

**Comment by danieldaeschle** [comment]
this is still failing

**Comment by JalonSolov** [comment]
Only failing on Mac, and only when compiling V UI.

**Comment by medvednikov** [comment]
Great work!
